Composite/Core/WebClient/Media/ImageResizer.cs

```diff
@@ -1,100 +1,139 @@
-﻿using System;
+using System;
 using System.Collections.Generic;
 using System.Drawing;
 using System.Drawing.Drawing2D;
 using System.Drawing.Imaging;
 using System.IO;
 using System.Linq;
+using System.Security.Cryptography;
+using System.Text;
 using System.Web;
 using System.Web.Caching;
+using System.Web.Hosting;
 using Composite.Core.IO;
 using Composite.Data.Plugins.DataProvider.Streams;
 using Composite.Data.Types;
 
 namespace Composite.Core.WebClient.Media
 {
     ///<summary>
-    ///Class that performs image resizing
+    /// Class that performs image resizing
     ///</summary>
     /// <exclude />
     [System.ComponentModel.EditorBrowsable(System.ComponentModel.EditorBrowsableState.Never)]
     public static class ImageResizer
     {
+        private static readonly MD5 _hashAlgorithm = MD5.Create();
+
         private const string ResizedImagesCacheDirectory = "~/App_Data/Composite/Cache/Resized images";
-        private static Dictionary<ImageFormat, string> _ImageFormat2Extension = new Dictionary<ImageFormat, string>
-        {  
-            {ImageFormat.Png, "png"},
-            {ImageFormat.Gif, "gif"}, 
-            {ImageFormat.Jpeg, "jpg"}, 
-            {ImageFormat.Bmp, "bmp"}, 
-            {ImageFormat.Tiff, "tiff"}
-        };
 
-        private static ImageCodecInfo JpegCodecInfo = ImageCodecInfo.GetImageEncoders().FirstOrDefault(codec => codec.FormatID == ImageFormat.Jpeg.Guid);
+        private static Dictionary<string, IImageFileFormatProvider> _imageFormatProviders;
+
+        private static Dictionary<string, IImageFileFormatProvider> ImageFormatProviders
+        {
+            get
+            {
+                if (_imageFormatProviders != null) return _imageFormatProviders;
+
+                var customProviders = ServiceLocator.GetServices<IImageFileFormatProvider>().ToList();

```

Comment:
Should we check that all services are already configured before saving _imageFormatProviders to the static property?